python/paddle/fluid/optimizer.py

```diff
@@ -2973,6 +2975,254 @@ def minimize(self,
+            step = 10
+
+            for i in range(step):
+                cost_val = exe.run(feed=gen_data(),
+                       program=fluid.default_main_program(),
+                       fetch_list=[cost.name])
+                print("step=%d cost=%f" % (i, cost_val[0]))
+
+    """
+
+    def __init__(self, optimizer, debug=False, debug_batchsize=32):
+        self._optimizer = optimizer
+        self._checkpoints = None
+        self.debug = debug
+        self.debug_batchsize = debug_batchsize
+
+    def _set_checkpoints(self, checkpoints):
+        self._checkpoints = checkpoints
+
+    def _get_var_size(self, block, name, batch):
+
+        dtype_to_size = {
+            core.VarDesc.VarType.FP16: 2,
+            core.VarDesc.VarType.FP32: 4,
+            core.VarDesc.VarType.FP64: 8,
+            core.VarDesc.VarType.INT16: 2,
+            core.VarDesc.VarType.INT32: 4,
+            core.VarDesc.VarType.INT64: 8,
+            core.VarDesc.VarType.BOOL: 1,
+            core.VarDesc.VarType.UINT8: 1,
+        }
+        if not block.desc.find_var(cpt.to_bytes(name)):
+            print("get var size failed, can not find var name %s" % name)
+            return 0
+        #print(block.desc.var(cpt.to_bytes(name)))
+        if block.desc.var(cpt.to_bytes(name)).type(
+        ) != core.VarDesc.VarType.LOD_TENSOR:
+            print("not lod tensor var, var name %s" % name)
+            return 0
+        var = block.var(name)
+        if var.shape[0] == -1:
+            res = -reduce(lambda x, y: x * y,
+                          var.shape) * batch * dtype_to_size[var.dtype]
+        else:
+            res = reduce(lambda x, y: x * y,
+                         var.shape) * dtype_to_size[var.dtype]
+        if res < 0:
+            return -res
+        return res
+
+    def draw(self, memory_with_position, recompute_segments):
+        #print("memory_with_position: ", memory_with_position)
+        #print("recompute_segments", recompute_segments)
+
+        import matplotlib
+        matplotlib.use('agg')
+        import matplotlib.pyplot as plt
+
+        x = [i - 2 for i in range(len(memory_with_position))]
+        y = [(i * 1.0) / 1024 / 1024 for i in memory_with_position]
+
+        if recompute_segments is not None:
+            for i in recompute_segments:
+                plt.axvline(x=i[1], color='r')
+
+        plt.plot(x, y, label="with_recompute")
+        #plt.plot(x2, y2, label="without_recompute") 
+
+        plt.legend(loc='upper right')
+        plt.xlabel("op_idx")
+        plt.ylabel('MB')
+        plt.title('')
+        plt.savefig('memory_anal.png')
+
+        import SimpleHTTPServer
+        import SocketServer
+
+        PORT = 8233
+
+        Handler = SimpleHTTPServer.SimpleHTTPRequestHandler
+
+        httpd = SocketServer.TCPServer(("", PORT), Handler)
+
+        #print("serving at port", PORT)
+        httpd.serve_forever()
+
+    def analysis_memory_usage(self, block):
```

Comment:
maybe you can make this a helper tool outside fluid python API. Once  this function is robust and exactly right w.r.t. GPU memory, you can make this function officially accessed by users.